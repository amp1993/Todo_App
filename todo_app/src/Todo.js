
// ***Todo***

// this component should display a ***div*** with the task of the todo.

// For each Todo component, there should also be a button with the text “X” that when clicked, removes the todo.

import React, { useState } from "react";

const Todo = ({ id, task, removeTodo, editTodo }) => {
    
    //set state for editing the task.Default will be false. 
    const [isEditing, setEditing] = useState(false);
    //Set state for when task has been edited and needs to update state in tasks. 
    const [editedTask, setEditedTask] = useState(task);

    //Handle edit button.
    const handleEdit = () => {
        setEditing(true);
    };

    //Save edited task that has been edited. 
    const handleSave = () => {
        editTodo(id, editedTask);
        setEditing(false);

    };

    return (
        //Have the remove and Edit button surface if isEditing is false. If isEditing is set to true, surface save button and hide remove and edit button. 
        <>
            
            {isEditing ? (
                <div id={id}>
                    <input
                        type="text"
                        value={editedTask}
                        onChange={(e) => setEditedTask(e.target.value)}
                    />
                    <button className="saveButton" onClick={() => { handleSave(id) }}>Save</button>
                </div>
            ) : (
                <div id={id}>
                    {task}
                    <button className="removeTodo" onClick={() => { removeTodo(id) }}>X</button>
                    <button className="editButton" onClick={() => { handleEdit(id, task) }}>Edit</button>
                </div>
            )}

        </>



    )

};

export default Todo;