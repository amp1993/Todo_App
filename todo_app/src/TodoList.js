

// ***TodoList***

// this component should render the ***NewTodoForm*** component and should render the list of Todo components. Place your state that contains all of the todos in this component.

import React, {useState} from "react";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";
import { v4 as uuidv4 } from 'uuid';


const TodoList  = () =>{

    const [todos, setTodos] = useState([]);

    const addTodo = (newTodo) =>{
        const id = uuidv4();
        setTodos(todos => [...todos, {...newTodo, id}])
    };

    const removeTodo = (id) =>{
        const updateTodos = todos.filter(todo => todo.id !== id);
        setTodos(updateTodos)
    };

    //Update Todos when it is edited. 
    const editTodo = (id, editedTask) => {
        const updatedTodos = todos.map((todo) =>
          todo.id === id ? { ...todo, task: editedTask } : todo
        );
        setTodos(updatedTodos);
      };

    return (
        <div className="todoList">
        <h2>Your Task List:</h2>
        <NewTodoForm addTodo={addTodo}   />
        {todos.map(({task, id})=> <Todo key={id} id={id} task={task} removeTodo={removeTodo} editTodo={editTodo}/>)}
        
        </div>
    )

} ;

export default TodoList;