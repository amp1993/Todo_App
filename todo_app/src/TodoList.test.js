import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";
import Todo from "./Todo";

it("renders without crashing", function() {
  render(<TodoList />);
});

it("matches snapshot", function() {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("can add a new todo", function () {
    const { getByLabelText, queryByText } = render(<TodoList />);
  
    // no todos yet
    expect(queryByText("Do Laundry")).not.toBeInTheDocument();
  
    // Use placeholder text for finding labels
    const todoInput = getByLabelText("Enter New Task:");
   
    const submitBtn = queryByText("Submit");
  
    // fill out the form
    fireEvent.change(todoInput, { target: { value: "Do Laundry" } });
    fireEvent.click(submitBtn);
  
    // item exists!
    expect(queryByText("Do Laundry")).toBeInTheDocument();
  }
  
  
  );



  test('removes todo when X button is clicked', () => {
    const todo = {
      id: 1,
      task: 'Example Todo',
    };
  
    const removeTodoMock = jest.fn();
    const { getByText } = render(<Todo {...todo} removeTodo={removeTodoMock} />);
  
    // Find the "X" button and click it
    const removeButton = getByText('X');
    fireEvent.click(removeButton);
  
    // Ensure that the removeTodo function is called with the correct ID
    expect(removeTodoMock).toHaveBeenCalledWith(todo.id);
  });