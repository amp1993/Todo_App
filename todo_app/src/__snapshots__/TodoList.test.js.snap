// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="todoList"
  >
    <h2>
      Your Task List:
    </h2>
    <form>
      <label
        for="task-input"
      >
        Enter New Task:
      </label>
      <input
        id="task-input"
        name="task"
        placeholder="task"
        type="text"
        value=""
      />
      <button>
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
